packages/ui/animations/useAnimatedList.tsx

```diff
@@ -0,0 +1,100 @@
+import { useMemo, useRef } from 'react';
+import { usePrevious } from '../hooks';
+
+interface UseAnimatedListProps<T = any> {
+   items: T[];
+   getId: (item: T) => string;
+}
+
+export function useAnimatedList<T>({ items, getId }: UseAnimatedListProps<T>) {
+   const animatedItems = useAnimatedItems({ items, getId });
+   const transitionItems = useTransitionItems(animatedItems);
+
+   const allItems = useMemo(() => {
+      const all = items.slice();
+      for (const indexedItem of transitionItems.exiting) {
+         all.splice(indexedItem.index, 0, indexedItem.item);
```

Comment:
Does this duplicate exiting items? I don't understand why this is needed

Reply:
`all` is a copy of the current `items` while `transition.exiting` contains items that are no longer present in `items`.
The purpose of this is to keep exiting items around for the time needed to the exit animations to complete (otherwise they'll be unmounted as soon as they're dropped from the array).

List animations in React are tricky because you have to keep exiting items around to be able to animate them, and this hook is accomplishing that in a transparent way.